### PR TITLE
Use async http library when interacting with s3

### LINKED ISSUE #602: Remove libcloud reliance for S3 compatible backends
[Project board link](https://github.com/orgs/k8ssandra/projects/8/views/1?pane=issue&itemId=34251642)

As we're only using a small subset of Libcloud's features and it's currently limiting our ability to make Medusa faster and safer, we have decided to move away from it and implement async http operations for multipart uploads.

In that effort, we will be moving the S3 and S3 compatible storage classes to rely on https://github.com/aiokitchen/aiohttp-s3-client, which relies on aiohttp and implements concurrent uploads of chunks during multipart uploads. It also handles retries in case of failure of single chunk uploads for resiliency to network glitches.

This also allows us to remove:
- The reliance on the awscli (and its dependencies)
- The concurrent code that was built to upload multiple files at the same time

We will be able to lower the default multipart threshold to 20 MB for better performance (benchmarks should verify this number is appropriate as multipart uploads may have an overhead).

Implementing rate limiting of uploads is out of scope for this ticket.

## Definition of Done
- [ ] Libcloud is not used anymore in the S3 storage classes
- [ ] ~All features that were previously supported in the Libcloud implementation must be ported over to the new one~
  - [ ] ~KMS encryption support~ (Moved to #612)
  - [ ] ~Multiple authentication capabilities (instance profiles, access key/secret, ...)~ (#613)
- [ ] ~Chunk size must be adjusted based on the total file size to stay within the 10k chunks limit in S3~ (#614)
- [ ] We do not implement concurrency in Medusa's code anymore
- [ ] We don't rely on the awscli for large files uploads
- [ ] The size threshold for multipart uploads is lowered at 20MB by default (anything bigger than that gets uploaded concurrently).

### PR DESCRIPTION
**Pull request by rzvoncek**
Fixes #602.

### CODE REVIEW

File: medusa/storage/s3_base_storage.py [hunk 1]
```diff
@@ -42,106 +53,359 @@
...
+        loop = self._get_or_create_event_loop()
+        b = loop.run_until_complete(self._read_blob_as_bytes(blob))
+        return b
+
+    async def _read_blob_as_bytes(self, blob: AbstractBlob) -> bytes:
+        async with self.http_client.get(blob.name) as resp:
+            return await resp.read()
+
+    # def list_objects(self, prefix: t.Optional[str] = None) -> t.List[AbstractBlob]:
+    #     return self.list_blobs(prefix)
```
Comment (adejanovski) [inline medusa/storage/s3_base_storage.py:187]: **todo:** if this isn't needed anymore, could you delete it?

**Review by adejanovski — APPROVED** [review]
Just a few minor changes to make and then you're good for a merge. Fantastic job!

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -105,8 +103,7 @@ jobs:
     strategy:
         fail-fast: false
         matrix:
-          #python-version: [3.6]
-          python-version: [3.6, "3.10"]
+          python-version: [3.9]
```
[the remark this answers is not in the document]
Comment (rzvoncek) [reply .github/workflows/ci.yml:106]: Created https://github.com/thelastpickle/cassandra-medusa/issues/620

File: medusa/cassandra_utils.py [hunk 3]
```diff
@@ -54,6 +54,9 @@ def list_files(self):
         # important to use the _r_glob() to recursively descend into subdirs if there are any
         return filter(lambda p: p.is_file(), self.path.rglob('*'))
 
+    def __repr__(self):
```
[the remark this answers is not in the document]
Comment (rzvoncek) [reply medusa/cassandra_utils.py:57]: This is called when an instance of the (SnapshotPath) class needs to be printed. It's almost like calling a `str()` on it. I'm not sure if the `str()` also calls the `__repr__`.

Without a repr, if we try to print or log the snapshot path, we get the internal calss identifier:
```
<medusa.cassandra_utils.SnapshotPath object at 0x10175b190>
```
This is what I spotted in a log somewhere.

So I added a repr and now we get:
```
SnapshotPath(path=/tmp, keyspace=k, table=t)
```

File: medusa/storage/s3_base_storage.py [hunk 4]
```diff
@@ -42,106 +53,359 @@
...
+        self.bucket_name: str = config.bucket_name
+
+        if config.storage_provider != 's3_compatible':
+            # assuming we're dealing with regular aws
+            s3_url = "https://{}.s3.amazonaws.com".format(self.bucket_name)
+        else:
+            # we're dealing with a custom s3 compatible storage, so we need to craft the URL
+            protocol = 'https' if config.secure.lower() == 'true' else 'http'
+            port = '' if config.port is None else str(config.port)
+            s3_url = '{}://{}:{}/{}'.format(protocol, config.host, port, self.bucket_name)
```
[the remark this answers is not in the document]
Comment (rzvoncek) [reply medusa/storage/s3_base_storage.py:131]: Created https://github.com/thelastpickle/cassandra-medusa/issues/621.

File: medusa/storage/s3_base_storage.py [hunk 5]
```diff
@@ -42,106 +53,359 @@
...
+
+
 class S3BaseStorage(AbstractStorage):
 
     def __init__(self, config):
-        self.session = botocore.session.Session()
 
+        self.api_version = '2006-03-01'
+
+        s3_region = 'us-west-2' if config.region == 'default' else config.region
```
[the remark this answers is not in the document]
Comment (rzvoncek) [reply medusa/storage/s3_base_storage.py]: Had to complement this by by setting the region in the config file used in the ITs (one of them at least). Our IT buckets are not in the default region it seems.

**Pull request merged**